projects/main_ping_delays_pingsender.kp/knowledge.md

```diff
@@ -0,0 +1,387 @@
+```
+
+```python
+p
+```
+
+
+
+
+    {'application/channel': u'nightly',
+     'creationDate': u'2017-04-02T22:39:42.529Z',
+     'id': u'5962c397-750e-4bbf-9bf2-839cfb0f272d',
+     'meta/Date': u'Mon, 03 Apr 2017 14:57:08 GMT',
+     'meta/Timestamp': 1491231459067853824L,
+     'meta/X-PingSender-Version': None,
+     'meta/creationTimestamp': 1.491172782529e+18,
+     'payload/info/reason': u'shutdown',
+     'payload/simpleMeasurements/shutdownDuration': 946,
+     'pre': 'pre'}
+
+
+
+We'll be plotting Cumulative Distribution Functions today.
+
+
+```python
+MAX_DELAY_S = 60 * 60 * 96.0
+HOUR_IN_S = 60 * 60.0
+PRES = ['pre', 'post']
+```
+
+```python
+def setup_plot(title, max_x, area_border_x=0.1, area_border_y=0.1):
+    plt.title(title)
+    plt.xlabel("Delay (hours)")
+    plt.ylabel("% of pings")
+
+    plt.xticks(range(0, int(max_x) + 1, 2))
+    plt.yticks(map(lambda y: y / 20.0, range(0, 21, 1)))
+
+    plt.ylim(0.0 - area_border_y, 1.0 + area_border_y)
+    plt.xlim(0.0 - area_border_x, max_x + area_border_x)
+
+    plt.grid(True)
+
+def plot_cdf(data):
+    sortd = np.sort(data)
+    ys = np.arange(len(sortd))/float(len(sortd))
+
+    plt.plot(sortd, ys)
+```
+
+```python
+def calculate_delays(p):
+    
+    created = datetime.fromtimestamp(p["meta/creationTimestamp"] / 1000.0 / 1000.0 / 1000.0)
+    received = datetime.fromtimestamp(p["meta/Timestamp"] / 1000.0 / 1000.0 / 1000.0)
+    sent = datetime.fromtimestamp(mktime_tz(parsedate_tz(p["meta/Date"]))) if p["meta/Date"] is not None else received
+    clock_skew = received - sent
+    
+    reporting_delay = (created - datetime.strptime(p["creationDate"], "%Y-%m-%dT%H:%M:%S.%fZ")).total_seconds()
+    submission_delay = (received - created - clock_skew).total_seconds()
+    return (reporting_delay, submission_delay)
+```
+
+```python
+delays_by_chan = combined_deduped.map(lambda p: (p["pre"], calculate_delays(p)))
+```
+### Submission Delay
+
+**Submission Delay** is the delay between the data being recorded on the client and it being received by our infrastructure. It is thought to be dominated by the length of time Firefox isn't open on a client's computer, though retransmission attempts and throttling can also contribute.
+
+
+```python
+setup_plot("'shutdown' ping submission delay CDF", MAX_DELAY_S / HOUR_IN_S, area_border_x=1.0)
+
+for pre in PRES:
+    plot_cdf(delays_by_chan\
+             .filter(lambda d: d[0] == pre)\
+             .map(lambda d: d[1][1] / HOUR_IN_S if d[1][1] < MAX_DELAY_S else MAX_DELAY_S / HOUR_IN_S)\
+             .collect())
+    
+plt.legend(["No pingsender", "With pingsender"], loc="lower right")
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7f3c05670750>
+
+
+
+
+
+![png](images/output_26_1.png)
+
+
+The use of `pingsender` results in an improvement in the submission delay of the `shutdown` "main" ping. We receive almost $85\%$ of the mentioned pings as soon as they are generated, instead of just ~$30\%$ within the first hour.
```

Comment:
`$85\%$` ...that's a weird thing there. A lot of your values in these paragraphs are surrounded by dollar signs.